x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map.ts

```diff
@@ -0,0 +1,173 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { PromiseReturnType } from '../../../typings/common';
+import { rangeFilter } from '../helpers/range_filter';
+import {
+  Setup,
+  SetupTimeRange,
+  SetupUIFilters
+} from '../helpers/setup_request';
+import { ENVIRONMENT_NOT_DEFINED } from '../../../common/environment_filter_values';
+import {
+  SERVICE_NAME,
+  SERVICE_ENVIRONMENT,
+  DESTINATION_ADDRESS,
+  CONNECTION_TYPE,
+  CONNECTION_SUBTYPE,
+  CALLEE_NAME,
+  CALLEE_ENVIRONMENT,
+  CONNECTION_UPSTREAM_LIST
+} from '../../../common/elasticsearch_fieldnames';
+
+export interface IEnvOptions {
+  setup: Setup & SetupTimeRange & SetupUIFilters;
+  serviceName?: string;
+  environment?: string;
+}
+
+interface ServiceMapElement {
+  data:
+    | {
+        id: string;
+      }
+    | {
+        id: string;
+        source: string;
+        target: string;
+      };
+}
+
+export type ServiceMapAPIResponse = PromiseReturnType<typeof getServiceMap>;
+export async function getServiceMap({
+  setup,
+  serviceName,
+  environment
+}: IEnvOptions) {
+  const { start, end, client, config, uiFiltersES } = setup;
+
+  const params = {
+    index: config['xpack.apm.serviceMapIndexPattern'],
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { range: rangeFilter(start, end) },
+            { exists: { field: CONNECTION_TYPE } },
+            ...uiFiltersES
+          ]
+        }
+      },
+      aggs: {
+        conns: {
+          composite: {
+            sources: [
+              { [SERVICE_NAME]: { terms: { field: SERVICE_NAME } } },
+              {
+                [SERVICE_ENVIRONMENT]: {
+                  terms: { field: SERVICE_ENVIRONMENT, missing_bucket: true }
+                }
+              },
+              {
+                [DESTINATION_ADDRESS]: {
+                  terms: {
+                    field: DESTINATION_ADDRESS
+                  }
+                }
+              },
+              { [CONNECTION_TYPE]: { terms: { field: CONNECTION_TYPE } } }, // will filter out regular spans etc.
+              {
+                [CONNECTION_SUBTYPE]: {
+                  terms: { field: CONNECTION_SUBTYPE, missing_bucket: true }
+                }
+              }
+            ],
+            size: 1000
+          },
+          aggs: {
+            dests: {
+              terms: {
+                field: CALLEE_NAME
+              },
+              aggs: {
+                envs: {
+                  terms: {
+                    field: CALLEE_ENVIRONMENT,
+                    missing: ''
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  if (serviceName || environment) {
+    const upstreamServiceName = serviceName || '*';
+    const upstreamEnvironment = environment
+      ? environment === ENVIRONMENT_NOT_DEFINED
+        ? 'null'
+        : environment
+      : '*';
```

Comment:
As we talked about perhaps it's better to remove the option to query for services without environment.

Reply:
The problem is that this is the global environment filter, which includes `Not defined` as an option. To remove it would affect other screens as well.

Reply:
but instead of using the string `'null'` to mark an environment which is not defined, i changed it to `ENVIRONMENT_NOT_DEFINED`, so that this nested ternary is no longer necessary.

Reply:
> To remove it would affect other screens as well.

yeah, I thought that's what we agreed - removing it everywhere? We should get 👍 from @formgeist but I don't see that being a problem.

Reply:
@sqren @ogupte I was under the impression early on that we'd have to remove the query bar, but I think that was even before we added the environment filter in there. Can you fill me in, what's the reason that we cannot have the "not defined" option in the environment filter in this case?